api/document/serializers/content.py

```diff
@@ -102,59 +102,91 @@ def to_internal_value(self, data: PrimitiveDict) -> PrimitiveDict:
         return serializer.to_internal_value(data)
 
 
-class BaseAnnotationSerializer(serializers.Serializer):
-    # This determines whether the annotation is a leaf of our
-    # content tree.
-    IS_LEAF = False
+class NestableAnnotationField(serializers.Field):
+    def get_attribute(self,
+                      instance: NestableAnnotation) -> NestableAnnotation:
+        return instance
 
-    content_type = serializers.SerializerMethodField()
-    inlines = serializers.SerializerMethodField()
-    text = serializers.SerializerMethodField()
 
-    @property
-    def CONTENT_TYPE(self):     # noqa; this is a constant
-        raise NotImplementedError('Unknown annotation type')
+class InlinesField(NestableAnnotationField):
+    def __init__(self, is_leaf_node: bool):
+        super().__init__()
+        self.is_leaf_node = is_leaf_node
+
+    def to_representation(self,
+                          instance: NestableAnnotation) -> List[PrimitiveDict]:
+        if self.is_leaf_node:
+            return []
+        return NestedAnnotationSerializer(
+            instance.children, context=self.context, many=True).data
+
+    def to_internal_value(self,
+                          data: List[PrimitiveDict]) -> List[PrimitiveDict]:
+        if not self.is_leaf_node:
+            serializer = NestedAnnotationSerializer()
+            return [
+                serializer.to_internal_value(item) for item in data
+            ]
+        elif data:
+            raise ValidationError('leaf nodes cannot contain nested content')
```

Comment:
"Leaf" seems like an internal term. What do you think about
```python
content_type = data.get('content_type', 'leaf')
raise ValidationError(f'{content_type} content can't contain nested content')
```

Reply:
Uhh, one annoying thing is that because `InlinesField`'s `to_internal_value()` is only being passed the value of `inlines` and not the parent object, um, I can't get ahold of that `content_type` unless I do some weird context gymnastics in the parent serializer.  Blaargh.  REST API is weird.